controller/src/main/java/io/pravega/controller/metrics/TransactionMetrics.java

```diff
@@ -29,11 +36,36 @@
 /**
  * Class to encapsulate the logic to report Controller service metrics for Transactions.
  */
-public final class TransactionMetrics extends AbstractControllerMetrics implements AutoCloseable {
+public final class TransactionMetrics extends AbstractControllerMetrics {
+
+    private static final AtomicReference<TransactionMetrics> INSTANCE = new AtomicReference<>();
+
+    private final OpStatsLogger createTransactionLatency;
+    private final OpStatsLogger createTransactionSegmentsLatency;
+    private final OpStatsLogger commitTransactionLatency;
+    private final OpStatsLogger commitTransactionSegmentsLatency;
+    private final OpStatsLogger committingTransactionLatency;
+    private final OpStatsLogger abortTransactionLatency;
+    private final OpStatsLogger abortTransactionSegmentsLatency;
+    private final OpStatsLogger abortingTransactionLatency;
 
-    private final OpStatsLogger createTransactionLatency = STATS_LOGGER.createStats(CREATE_TRANSACTION_LATENCY);
-    private final OpStatsLogger commitTransactionLatency = STATS_LOGGER.createStats(COMMIT_TRANSACTION_LATENCY);
-    private final OpStatsLogger abortTransactionLatency = STATS_LOGGER.createStats(ABORT_TRANSACTION_LATENCY);
+    private TransactionMetrics() {
+        createTransactionLatency = STATS_LOGGER.createStats(CREATE_TRANSACTION_LATENCY);
+        createTransactionSegmentsLatency = STATS_LOGGER.createStats(CREATE_TRANSACTION_SEGMENTS_LATENCY);
+        commitTransactionLatency = STATS_LOGGER.createStats(COMMIT_TRANSACTION_LATENCY);
+        commitTransactionSegmentsLatency = STATS_LOGGER.createStats(COMMIT_TRANSACTION_SEGMENTS_LATENCY);
+        committingTransactionLatency = STATS_LOGGER.createStats(COMMITTING_TRANSACTION_LATENCY);
+        abortTransactionLatency = STATS_LOGGER.createStats(ABORT_TRANSACTION_LATENCY);
+        abortTransactionSegmentsLatency = STATS_LOGGER.createStats(ABORT_TRANSACTION_SEGMENTS_LATENCY);
+        abortingTransactionLatency = STATS_LOGGER.createStats(ABORTING_TRANSACTION_LATENCY);
+    }
+
+    public static synchronized TransactionMetrics getInstance() {
+        if (INSTANCE.get() == null) {
```

Comment:
A `getInstance()` call post `reset()` will create a new instance of TransactionMetrics. This can be avoided.
(Using AutoCloseable interface is also desirable).

Reply:
the intent of reset is to destroy the previous instantiation. 

Also, auto closable will be on the object. Which would mean the object could be closed and getInstance could return a closed object. Which should not be allowed. 
The close method on the object should be effectively private (logically it is similar to what @RaulGracia has done now).

Reply:
ok.
I assumed `reset()` is invoked when we are planning to close the instance of TransactionMetrics i.e. after reset() is invoked we would not want to invoke `getInstance()`. I was thinking more on the lines where calling `getInstance()` post the close would throw an `ObjectClosedException`.